Tutorial on creating an inference snap (#80)
canonical/inference-snaps: docs/tutorial/create-inference-snap.md

## What you'll need

We use `snapcraft` to create snaps.
If you don't have it installed, refer to [this guide](https://documentation.ubuntu.com/snapcraft/stable/how-to/set-up-snapcraft/)

You will also need the model weights and a compatible runtime. 
But first, we need to pick an LLM. Gemma 3 is a good choice for this tutorial because it is open and available in small sizes.
Gemma 3 is free and permits responsible commercial use.
Read its terms of use [here](https://ai.google.dev/gemma/terms).

Inferencing with a CPU is always a good starting point and a simple way to do that is via {spellexception}`llama.cpp` runtime. 
The {spellexception}`llama.cpp` project is backed by a large open source community working on various silicon optimizations. 
We can use the llama-server implementation which supports model weights in the GGUF format.
Google distributes GGUF formats for most of Gemma 3 model sizes. 
Go with a relatively small model to speed up development times and add larger ones later.

The [`gemma-3-1b-it-qat-q4_0-gguf`](https://huggingface.co/google/gemma-3-1b-it-qat-q4_0-gguf) is a good first instance.
Download the `.gguf` file.

> 

Author: This is covered in later sections. The prep section is just about collecting everything that is needed to start. Let me know if you think it isn't clear.

# Your first inference snap

In this tutorial we will create an inference snap for a large language model (LLM).
The focus is on learning the overall structure by packaging a single model and runtime, rather than making a snap that works on a wide range of machines.

> Once we learn the basics, it is easy to extend it to include other model weights and runtimes. 

Reviewer: Same here and globally; also, don't overpromise:
```suggestion
Once you learn the basics, you'll be able to apply the newly gained skill for other model weights and runtimes. 
```

You will also need the model weights and a compatible runtime. 
But first, we need to pick an LLM. Gemma 3 is a good choice for this tutorial because it is open and available in small sizes.
Gemma 3 is free and permits responsible commercial use.
Read its full terms of use [here](https://ai.google.dev/gemma/terms).

Inference with a CPU is always a good starting point and a simple way to do that is via a {spellexception}`llama.cpp`-based runtime. 
The [{spellexception}`llama.cpp`](https://github.com/ggml-org/llama.cpp) project is backed by a large open source community working on various silicon optimizations. 
We can use the [{spellexception}`llama.cpp` HTTP server](https://github.com/ggml-org/llama.cpp/tree/master/tools/server) implementation which supports model weights in the GGUF format.

Google distributes GGUF formats for most of Gemma 3 model sizes. 
Go with a relatively small model to speed up development times and add larger ones later.
The [`gemma-3-1b-it-qat-q4_0-gguf`](https://huggingface.co/google/gemma-3-1b-it-qat-q4_0-gguf) is a good first instance.

> Download the `gemma-3-1b-it-q4_0.gguf` file from the Hugging Face repository.

Reviewer: Rework to make it clear the link points to the repo in question:

```suggestion
A good first model to use is the [`gemma-3-1b-it-qat-q4_0-gguf`](https://huggingface.co/google/gemma-3-1b-it-qat-q4_0-gguf) variant on Hugging Face.  
On that page, download the {file}`gemma-3-1b-it-q4_0.gguf` file (the file name differs slightly from the model name).
```

We need to specify what gets packaged in them.

Add a new *part* to copy the local files from the `components` directory to the corresponding component within the package.

```{literalinclude} create-inference-snap/snap/snapcraft.yaml
:language: yaml
:start-after: start component-local-files part
:end-before: end part
```

`(component/<component name>)` is a special syntax to reference a defined component.

Furthermore, add a *part* to build {spellexception}`llama.cpp` from source and move the artifacts to the corresponding component:
```{literalinclude} create-inference-snap/snap/snapcraft.yaml
:language: yaml
:start-after: start llama-cpp part
:end-before: end part
```

We don't need all the {spellexception}`llama.cpp` artifacts, but that's okay for now. 
You can refine this *part* later to only include what is necessary.

The packaging logic for the components are now complete.
You can build the snap and verify the packages:
```{terminal}
:input: snapcraft pack
Packed: gemma3-jane_v3_amd64.snap, gemma3-jane+model-1b-it-q4-0-gguf.comp, gemma3-jane+llama-cpp.comp
```

### Engine

The runtime and model weights are tightly coupled components useful for certain host machines.

> Each inference snaps can include a range of such components. 

Reviewer: ```suggestion
Each inference snap can include a range of such components. 
```

In this tutorial we create a single {ref}`engine <engines>` that uses those components. 
Engines can reuse and customize components, for example to bundle a runtime with different model weights or to tweak the command line arguments.

Create an `engines` directory, with a subdirectory for our only engine named `generic-cpu`. 

Inside, create an `engine.yaml` file that defines the engine:
```{literalinclude} create-inference-snap/engines/generic-cpu/engine.yaml
:caption: engines/generic-cpu/engine.yaml
:language: yaml
```

> That is the {ref}`engine manifest file <engine-manifest>`. 

Reviewer: Good decision, but be cautious: too many links in a tutorial could create a non-linear narrative that troubles the beginner.

The reason for this is that the snap does not have access to detect the hardware and select the engine.

Grant the snap access to the `hardware-observe` interface:
```shell
sudo snap connect gemma3-jane:hardware-observe
```

Now, request auto selection of the engine:
```{terminal}
:input: sudo gemma3-jane use-engine --auto
Evaluating engines for optimal hardware compatibility:
✔ generic-cpu: compatible, score=12
Selected engine: generic-cpu
Engine successfully changed to "generic-cpu"
```
As the output shows, the snap has detected that `generic-cpu` engine is compatible with the host machine. 
If there were other engines, it would have picked the best matching one.

Start the server:
```shell
sudo snap start gemma3-jane
```

After starting the server, check the logs again and make sure it started successfully:
```{terminal}
:input: sudo snap logs gemma3-jane
...
2025-11-13T15:34:52+01:00 gemma3-jane.server[241059]: main: server is listening on http://127.0.0.1:9090 - starting the main loop
2025-11-13T15:34:52+01:00 gemma3-jane.server[241059]: srv  update_slots: all slots are idle
```

Looks like the server is up and running!

> Submit a request via curl:

Reviewer: ```suggestion
Submit a request using {command}`curl`:
```

    "max_tokens": 30,
    "temperature": 0,
    "stream": false,
    "messages": [
      { "role": "system", "content": "You are a helpful assistant." },
      { "role": "user", "content": "What are the 3 main tourist attractions in Paris?" }
    ]                                                                                   
  }'
{"choices":[{"finish_reason":"length","index":0,"message":{"role":"assistant","content":"Okay, here are the 3 main tourist attractions in Paris:\n\n1.  **The Eiffel Tower:** Undoubtedly the most iconic landmark in the world"}}],"created":1763044754,"model":"gpt-3.5-turbo","system_fingerprint":"b1-6a746cf","object":"chat.completion","usage":{"completion_tokens":30,"prompt_tokens":28,"total_tokens":58},"id":"chatcmpl-PLT9wXbZiJ6XXxHCQUBsTz9wj59y2KHU","timings":{"prompt_n":28,"prompt_ms":90.995,"prompt_per_token_ms":3.249821428571429,"prompt_per_second":307.7092147920215,"predicted_n":30,"predicted_ms":833.198,"predicted_per_token_ms":27.773266666666665,"predicted_per_second":36.005847349609574}}
```

That worked! We have created an inference snap for Gemma 3 model, with just one engine.

> ````{tip}

Reviewer: The tip is too bulky; I suggest including it in the main text.

The {spellexception}`llama.cpp` engine we created also includes a chat web GUI.
Access it via the port as the API: http://localhost:9090

![Web GUI screenshot](create-inference-snap/gui-screenshot.png)

````

### Publish to the store

Uploading the snap to the store unlocks the automatic engine selection and component installation functionality.

> Moreover, it lets others easily install and use the snap.

Reviewer: ```suggestion
Moreover, it lets others install and use the snap.
```

One's 'easily' is another one's nightmare. Same for 'simply', etc.

> Uploading requires a few steps. Here is the list:

Reviewer: It's better to rework this into appropriately titled subsections rather than listing all actions and then writing a single block of text.

Inference with a CPU is always a good starting point and a simple way to do that is via a {spellexception}`llama.cpp`-based runtime. 
The [{spellexception}`llama.cpp`](https://github.com/ggml-org/llama.cpp) project is backed by a large open source community working on various silicon optimizations. 
We can use the [{spellexception}`llama.cpp` HTTP server](https://github.com/ggml-org/llama.cpp/tree/master/tools/server) implementation which supports model weights in the GGUF format.

Google distributes GGUF formats for most of Gemma 3 model sizes. 
Go with a relatively small model to speed up development times and add larger ones later.
The [`gemma-3-1b-it-qat-q4_0-gguf`](https://huggingface.co/google/gemma-3-1b-it-qat-q4_0-gguf) is a good first instance.
Download the `gemma-3-1b-it-q4_0.gguf` file from the Hugging Face repository.

You may also use `tree` and `curl` which you can install with:
```shell
sudo apt update
sudo apt install tree curl
```

### The snapcraft.yaml

The snapcraft YAML file is where the snap's packaging logic is defined.
Use the following command to create a template in an initial project tree:
```shell

> snapcraft init

Reviewer 2: If a user blindly follows this, it will create a new snapcraft.yaml file inside the `~/home/snap` directory. How about being more explicit with a snap name.
```suggestion
snapcraft init my-inference-snap
```

```

Checkout the generated file:
```{terminal}
:input: tree .
.
└── snap
    └── snapcraft.yaml

2 directories, 1 file
```

Update the snapcraft file with sensible metadata.
Remove the parts that were generated by default.

Inference snaps are usually named after the model (e.g., `gemma3` for Gemma 3 model). 

> Add your username to it as a suffix to create a unique snap name that can be uploaded to the store in order to complete this tutorial.

Author: ```suggestion
Add your username to it as a suffix to create a unique snap name that can be uploaded to the store in order to complete this tutorial. In this tutorial, `jane` is used as a placeholder for your username.
```

The snap store can grant the snap permission to auto connect the hardware-observe interface. This requires a manual review by the store security team.

Submit a request on the [Snapcraft forum](https://forum.snapcraft.io/c/store-requests/privileged-interfaces/27) to get this permission for your snap.
An example, extended request can be found [here](https://forum.snapcraft.io/t/auto-connection-of-hardware-observe-and-home-for-gemma3/49246).

Once the permission is granted, you can install the snap in confined mode, without setting `--devmode` and it will be ready to use right away.

## Next steps

We created a basic inference snap for Gemma 3 model with a single engine. 
You can extend it in various ways, for example to add an additional engine for GPU support.
This could be another build of {spellexception}`llama.cpp` targeting CUDA, or a different runtime altogether, such as [OpenVINO Model Server](https://github.com/openvinotoolkit/model_server) which supports Intel GPUs and NPUs.

You may also have noticed the CLI doesn't offer tab completion.
You can find some guidance for adding it temporarily, e.g. for Bash bu running `gemma3-jane completion bash -h`.

> Ideally, the tab completion should be added to the snap itself so becomes available during the installation.

Reviewer 2: ```suggestion
Ideally, the tab completion should be added to the snap itself so it becomes available during the installation.
```
